kornia/utils/draw.py

```diff
@@ -47,87 +47,99 @@ def draw_line(image: torch.Tensor, p1: torch.Tensor, p2: torch.Tensor, color: to
-    elif -1 < m < 0:
-        for i in range(x1, x2 + 1):
-            _draw_pixel(image, i, y2, color)
-            if line_equation(i + 1, y2 - 0.5) > 0:
-                y2 -= 1
-
+    x1, y1 = p1[..., 0], p1[..., 1]
+    x2, y2 = p2[..., 0], p2[..., 1]
+    dx = x2 - x1
+    dy = y2 - y1
+    dx_sign = torch.sign(dx)
+    dy_sign = torch.sign(dy)
+    dx, dy = torch.abs(dx), torch.abs(dy)
+    dx_zero_mask = dx == 0
+    dy_zero_mask = dy == 0
+    dx_gt_dy_mask = (dx > dy) & ~(dx_zero_mask | dy_zero_mask)
+    rest_mask = ~(dx_zero_mask | dy_zero_mask | dx_gt_dy_mask)
+
+    dx_zero_x_cords, dx_zero_y_cords = [], []
+    dy_zero_x_cords, dy_zero_y_cords = [], []
+    dx_gt_dy_x_cords, dx_gt_dy_y_cords = [], []
+    rest_x_cords, rest_y_cords = [], []
+
+    if dx_zero_mask.any():
+        dx_zero_x_cords = [
+            x for x_i, dy_i in zip(x1[dx_zero_mask], dy[dx_zero_mask]) for x in x_i.repeat(int(dy_i.item() + 1))
+        ]
+        dx_zero_y_cords = [
+            y
+            for y_i, s, dy_ in zip(y1[dx_zero_mask], dy_sign[dx_zero_mask], dy[dx_zero_mask])
+            for y in (y_i + s * torch.arange(0, dy_ + 1, 1).to(image.device))
+        ]
+    if dy_zero_mask.any():
+        dy_zero_x_cords = [
+            x
+            for x_i, s, dx_i in zip(x1[dy_zero_mask], dx_sign[dy_zero_mask], dx[dy_zero_mask])
+            for x in (x_i + s * torch.arange(0, dx_i + 1, 1).to(image.device))
+        ]
+        dy_zero_y_cords = [
+            y for y_i, dx_i in zip(y1[dy_zero_mask], dx[dy_zero_mask]) for y in y_i.repeat(int(dx_i.item() + 1))
+        ]
+    if dx_gt_dy_mask.any():
+        dx_gt_dy_x_cords = [
+            x
+            for x_i, s, dx_i in zip(x1[dx_gt_dy_mask], dx_sign[dx_gt_dy_mask], dx[dx_gt_dy_mask])
+            for x in (x_i + s * torch.arange(0, dx_i + 1, 1).to(image.device))
+        ]
+        dx_gt_dy_y_cords = [
+            y
+            for y_i, s, dx_i, dy_i in zip(
+                y1[dx_gt_dy_mask], dy_sign[dx_gt_dy_mask], dx[dx_gt_dy_mask], dy[dx_gt_dy_mask]
+            )
+            for y in (y_i + s * torch.arange(0, dy_i + 1, dy_i / dx_i)[: int(dx_i.item()) + 1].ceil().to(image.device))
+        ]
+    if rest_mask.any():
+        rest_x_cords = [
+            x
+            for x_i, s, dx_i, dy_ in zip(x1[rest_mask], dx_sign[rest_mask], dx[rest_mask], dy[rest_mask])
+            for x in (x_i + s * torch.arange(0, dx_i + 1, dx_i / dy_)[: int(dy_.item()) + 1].ceil().to(image.device))
+        ]
+        rest_y_cords = [
+            y
+            for y_i, s, dy_i in zip(y1[rest_mask], dy_sign[rest_mask], dy[rest_mask])
+            for y in (y_i + s * torch.arange(0, dy_i + 1, 1).to(image.device))
+        ]
+    x_cords = torch.tensor(dx_zero_x_cords + dy_zero_x_cords + dx_gt_dy_x_cords + rest_x_cords).long()
+    y_cords = torch.tensor(dx_zero_y_cords + dy_zero_y_cords + dx_gt_dy_y_cords + rest_y_cords).long()
+    image[:, y_cords, x_cords] = color.view(-1, 1)
```

Comment:
Just to think in next steps, how this method will support adjusting the line with; or giving subpixel coordinates?

Reply:
I missed the width and sub-pixel parts. Will work on that

Reply:
Should we merge this as it is, and work on this in a subsequent PR?